storage/kafka/datasource.go

```diff
@@ -0,0 +1,349 @@
+	"github.com/cube2222/octosql/physical/metadata"
+	"github.com/cube2222/octosql/streaming/storage"
+)
+
+var availableFilters = map[physical.FieldType]map[physical.Relation]struct{}{
+	physical.Primary:   {},
+	physical.Secondary: {},
+}
+
+type DataSource struct {
+	brokers      []string
+	topic        string
+	partition    int
+	batchSize    int
+	decodeAsJSON bool
+	alias        string
+	stateStorage storage.Storage
+}
+
+func NewDataSourceBuilderFactory(partitions int) physical.DataSourceBuilderFactory {
+	return physical.NewDataSourceBuilderFactory(
+		func(ctx context.Context, matCtx *physical.MaterializationContext, dbConfig map[string]interface{}, filter physical.Formula, alias string, partition int) (execution.Node, error) {
+			// Get execution configuration
+			hosts, ports, err := config.GetIPAddressList(dbConfig, "brokers", config.WithDefault([]interface{}{[]string{"localhost"}, []int{9092}}))
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get address")
+			}
+			topic, err := config.GetString(dbConfig, "topic")
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get topic")
+			}
+			batchSize, err := config.GetInt(dbConfig, "batchSize", config.WithDefault(1))
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get batch size")
+			}
+			decodeAsJSON, err := config.GetBool(dbConfig, "json", config.WithDefault(false))
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get json option")
+			}
+
+			brokers := make([]string, len(hosts))
+			for i := range hosts {
+				brokers[i] = fmt.Sprintf("%s:%d", hosts[i], ports[i])
+			}
+
+			return &DataSource{
+				brokers:      brokers,
+				topic:        topic,
+				partition:    partition,
+				batchSize:    batchSize,
+				decodeAsJSON: decodeAsJSON,
+				alias:        alias,
+				stateStorage: matCtx.Storage,
+			}, nil
+		},
+		nil,
+		availableFilters,
+		metadata.Unbounded,
+		partitions,
+	)
+}
+
+// NewDataSourceBuilderFactoryFromConfig creates a data source builder factory using the configuration.
+func NewDataSourceBuilderFactoryFromConfig(dbConfig map[string]interface{}) (physical.DataSourceBuilderFactory, error) {
+	partitions, err := config.GetInt(dbConfig, "partitions", config.WithDefault(1))
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get primaryKeys")
+	}
+
+	return NewDataSourceBuilderFactory(partitions), nil
+}
+
+var offsetPrefix = []byte("$kafka_offset$")
+var tokenQueuePrefix = []byte("$token_queue$")
+
+func (ds *DataSource) Get(ctx context.Context, variables octosql.Variables, streamID *execution.StreamID) (execution.RecordStream, error) {
+	tx := storage.GetStateTransactionFromContext(ctx).WithPrefix(streamID.AsPrefix())
+
+	dialer := &kafka.Dialer{
+		Timeout:   10 * time.Second,
+		DualStack: true,
+	}
+
+	r := kafka.NewReader(kafka.ReaderConfig{
+		Brokers:   ds.brokers,
+		Topic:     ds.topic,
+		Dialer:    dialer,
+		Partition: ds.partition,
+		MinBytes:  10e1,
+		MaxBytes:  10e7,
```

Comment:
Imho we could extract those values into some named constants

Reply:
I'm not for that. Because they're used in only this one location, with the descriptive variable next to them. It's like a const client config really. The variable name won't get any more descriptive. Those are some values I've chosen in the past because they worked for me, others would probably work well too.